src/libAtomVM/otp_socket.c

```diff
@@ -1069,62 +1084,100 @@ static term nif_socket_setopt(Context *ctx, int argc, term argv[])
+                term linger = interop_kv_get_value(value, linger_atom, ctx->global);
+                VALIDATE_VALUE(linger, term_is_integer);
 
 #if OTP_SOCKET_BSD
-        struct linger sl;
-        sl.l_onoff = (onoff == TRUE_ATOM);
-        sl.l_linger = term_to_int(linger);
-        int res = setsockopt(rsrc_obj->fd, SOL_SOCKET, SO_LINGER, &sl, sizeof(sl));
-        if (UNLIKELY(res != 0)) {
-            return make_errno_tuple(ctx);
-        } else {
-            return OK_ATOM;
-        }
+                struct linger sl;
+                sl.l_onoff = (onoff == TRUE_ATOM);
+                sl.l_linger = term_to_int(linger);
+                int res = setsockopt(rsrc_obj->fd, SOL_SOCKET, SO_LINGER, &sl, sizeof(sl));
+                if (UNLIKELY(res != 0)) {
+                    return make_errno_tuple(ctx);
+                } else {
+                    return OK_ATOM;
+                }
 #elif OTP_SOCKET_LWIP
-        rsrc_obj->linger_on = (onoff == TRUE_ATOM);
-        rsrc_obj->linger_sec = term_to_int(linger);
-        return OK_ATOM;
+                rsrc_obj->linger_on = (onoff == TRUE_ATOM);
+                rsrc_obj->linger_sec = term_to_int(linger);
+                return OK_ATOM;
 #endif
-        // TODO add more as needed
-        // int flag = 1;
-        // int res = setsockopt(fd, IPPROTO_TCP, TCP_NODELAY, (char *) &flag, sizeof(int));
-        // if (UNLIKELY(res != 0)) {
-        //     AVM_LOGW(TAG, "Failed to set TCP_NODELAY.");
-        // }
-    } else {
-        RAISE_ERROR(BADARG_ATOM);
+                // TODO add more as needed
+                // int flag = 1;
+                // int res = setsockopt(fd, IPPROTO_TCP, TCP_NODELAY, (char *) &flag, sizeof(int));
+                // if (UNLIKELY(res != 0)) {
+                //     AVM_LOGW(TAG, "Failed to set TCP_NODELAY.");
+                // }
+            } else {
+                RAISE_ERROR(BADARG_ATOM);
+            }
+
+            case OtpSocketSetoptLevelOTP: {
+                term opt = term_get_tuple_element(level_tuple, 1);
+                if (globalcontext_is_term_equal_to_atom_string(global, opt, rcvbuf_atom)) {
+                    // socket:setopt(Socket, {otp, recvbuf}, BufSize :: non_neg_integer())
+
+                    if (UNLIKELY(!term_is_integer(value))) {
+                        AVM_LOGE(TAG, "socket:setopt: otp rcvbuf value must be an integer");
+                        return make_error_tuple(globalcontext_make_atom(global, invalid_value_atom), ctx);
+                    }
+
+                    avm_int_t buf_size = term_to_int(value);
+                    if (UNLIKELY(buf_size < 0)) {
+                        AVM_LOGE(TAG, "socket:setopt: otp rcvbuf value may not be negative");
+                        return make_error_tuple(globalcontext_make_atom(global, invalid_value_atom), ctx);
+                    }
+
+                    rsrc_obj->buf_size = (size_t) buf_size;
```

Comment:
We should call setsockopt here instead.
Idea is this option is here to limit the amount of memory, so let's tell the IP stack what we want.

Reply:
I am a little confused here.  This is setting the buffer size we allocate at the OTP layer — I.e the size of the binary (in the case of peek) or the size of the malloc (in the case of no peek), not at the IP layer.  Or are you saying we should use the IP layer as the authoritative value?

Reply:
Sorry for bringing my own confusion in.

Indeed, I was suggesting that we use the IP Layer buffer size.
But in Erlang/OTP, there actually are `{socket, rcvbuf}` and `{otp, rcvbuf}`.

So maybe please disregard my comments and move forward with the implementation of `{otp, rcvbuf}`. My comment on the default needing to be at least as large as the MTU for UDP still applies, though, so maybe it should be 1500 (or 1460).

We also may want to implement `{socket, rcvbuf}`.